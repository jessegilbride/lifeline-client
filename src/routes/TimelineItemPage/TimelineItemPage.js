import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import TimelineContext from '../../contexts/TimelineContext';
import TimelineApiService from '../../services/timeline-api-service';
import { NiceDate } from '../../components/Utils/Utils';
import './TimelineItemPage.css';

export default class TimelineItemPage extends Component {
  static defaultProps = {
    match: { params: {} } // React Router default
  };

  static contextType = TimelineContext;

  state = { 
    currentItem: {},
    error: null
  }

  componentDidMount() {
    const { itemId } = this.props.match.params; // React Router prop
    // const { items } = this.context;

    TimelineApiService
      .getItemById(itemId)
      .then(item => {
        /* if (Date.parse(item.entry_date).isNaN) {
          console.log('not a number')
        } else {
          console.log(Date.parse(item.entry_date))
        } */
        this.setState({currentItem: item})
      })
      .catch(error => {
        console.log(error)
        this.setState({error})
      });
  }

  makeTimelineItemPage() {
    // const { line } = this.context;

    const item = this.state.currentItem;

    /* function ComponentName() {
      return <div className=''></div>;
    } <ComponentName /> */

    return (
      <>
        <header>
          <h2>{item.title}</h2>
        </header>
        <p>
          <NiceDate date={item.entry_date} />
        </p>
        <p>
          {item.content}
        </p>
      </>
    );
  }

  render() {
    const { currentItem, error } = this.state;
    let pageContent;
    if (error) {
      console.log('TimelineItemPage render error: ', error)
      pageContent =
        // This error 'string' must match the error response from items-router.checkItemExists
        error.error === `Item does not exist.` ? (
          <p className='red'>Item does not exist.</p>
        ) : (
          <p className='red'>There was an error. Sorry about that. Head <Link to='timeline'>back to the timeline</Link>?</p>
        );
    } else if (!currentItem) { // << fix?
      pageContent = <span className='spinner-container'><div className="lds-roller"> <div></div> <div></div> <div></div> </div></span>;
    } else {
      pageContent = this.makeTimelineItemPage();
    }
    return <section className='LineItemView'>{pageContent}</section>;
  }

}
