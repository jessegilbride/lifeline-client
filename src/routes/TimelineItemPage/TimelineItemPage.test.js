import React from 'react';
import ReactDOM from 'react-dom';
import { MemoryRouter } from "react-router-dom";
import renderer from 'react-test-renderer';
import TimelineItemPage from './TimelineItemPage';

describe('TimelineItemPage component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
    <MemoryRouter><TimelineItemPage /></MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    const tree = renderer
      .create(
      <MemoryRouter><TimelineItemPage name="TimelineItemPage" /></MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });

});
