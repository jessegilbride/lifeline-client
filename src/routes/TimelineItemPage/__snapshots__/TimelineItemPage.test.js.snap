// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineItemPage component renders the UI as expected 1`] = `
<section
  className="LineItemView"
>
  <header>
    <h2 />
  </header>
  <p>
    Invalid Date
  </p>
  <p />
</section>
`;
